File: Original code from paper/KinDrape.m
function [Node,P] = KinDrape(d,Grid,Org,Ang,OrgNode)
%% Mold definition: Hemisphere
[Theta,Phi] = meshgrid(linspace(0,2*pi,100),linspace(pi/20,pi/2-1e-10,50));
[X,Y,Z] = sph2cart(Theta,Phi,1); 
F = scatteredInterpolant(X(:),Y(:),Z(:),'linear','linear');
%% Auxiliary variables, solver settings and initialization of Node and P
Dir1 = [1 0 ; 0 1 ; -1 0 ; 0 -1]; Dir2 = [-Dir1(:,2) Dir1(:,1)];
Opt1 = optimoptions(@fsolve,'Display','off'); 
Opt2 = optimoptions(@fmincon,'Algorithm','active-set','Display','notify'...
    ,'MaxFunctionEvaluations',5e3);
Node = NaN([Grid 3]); P = NaN(prod(Grid-1),4,4);
%% Step 1: Place org. node (1) and node (2) defined by ini. drape angle
% Define linear indices for cell, place 1st node and solve for 2nd node
Idx = CellIdx(Grid,OrgNode(1),OrgNode(2),Dir1,Dir2,1);
Node(Idx(1,:)) = [Org(1), Org(2), F(Org(1), Org(2))]; 
a_sol = fsolve(@(a)DistFun(a,Node(Idx(1,:)),F,d,Ang),3/4*d,Opt1);
Node(Idx(1:2,:)) = CellVertCoor(a_sol,Node(Idx(1,:)),F,Ang);
%% Step 2: Place generator cells (initial cells) while minimizing shear
GenStart = OrgNode + [0 0 ; 1 1 ; 0 1 ; 0 0];
nGenCell = [Grid-OrgNode-[0 1]  OrgNode-1]; 
for i = 1:4
    a_0 = repmat(3/4*d*[cosd(Ang+(i-1)*90) sind(Ang+(i-1)*90)],1,2);
    for j = GenStart(i,:)' + (0:nGenCell(i)-1).*Dir1(i,:)'
        % Get cell idx and def. solver input. Call fmincon, assign solution
        [Idx, CellNo] = CellIdx(Grid,j(1),j(2),Dir1,Dir2,i);
        Bnd = a_0 + 1/2*norm(a_0(1:2))*[-1 -1 -1 -1; 1 1 1 1];
        a_sol = fmincon(@(a)ShearFun(a,Node(Idx),F),a_0,[],[],[],[],...
            Bnd(1,:),Bnd(2,:),@(a)DistFun(a,Node(Idx),F,d,[]),Opt2);
        [~,Node(Idx),Shear] = ShearFun(a_sol,Node(Idx),F);
        % Put current cell coord. and shear in P array and update a_0
        P(CellNo,1:4,1:4) = [Node(Idx) Shear']; 
        a_0 = a_sol;
    end
end
%% Step 3: Place remaining, constrained cells
ConStart = OrgNode + [1 1 ; 0 1 ; 0 0 ; 1 0];
nConCell = nGenCell([1 2 ; 3 2 ; 3 4 ; 1 4]) - [1 0 ; 0 0 ; 0 0 ; 1 0];
for i = 1:4
    a_0 = 3/4*d*[cosd(Ang+(i-1)*90) sind(Ang+(i-1)*90)];
    for j = ConStart(i,1) + (0:nConCell(i,1)-1)*(Dir1(i,1)+Dir2(i,1))
        for k = ConStart(i,2) + (0:nConCell(i,2)-1)*(Dir1(i,2)+Dir2(i,2)) 
            % Get cell idx and def. solver input. Call fsolve, assign sol.
            [Idx, CellNo] = CellIdx(Grid,j,k,Dir1,Dir2,i);
            a_sol = fsolve(@(a)DistFun(a,Node(Idx),F,d,[]),a_0,Opt1);
            [~,Node(Idx),Shear] = ShearFun(a_sol,Node(Idx),F);
            % Put current cell coord. and shear in P array and update a_0
            P(CellNo,1:4,1:4) = [Node(Idx) Shear']; 
            a_0 = a_sol;
        end
    end
end
%% Plot
figure; scatter3(Org(1),Org(2),F(Org(1),Org(2)),'kx','LineWidth',5); 
hold on; axis('equal','tight'); xlabel('x'); ylabel('y'); zlabel('z');
surf(X,Y,Z,0,'EdgeColor','none','FaceColor',[0.6,0.7,0.8],'FaceAlpha',0.5);
patch(P(:,:,1)',P(:,:,2)',P(:,:,3)',P(:,:,4)');
cb = colorbar; cb.Label.String = 'Shear Angle [deg]'; colormap('jet');
end
%% Aux. functions
function [Idx, CellNo] = CellIdx(Grid,Row,Col,Dir1,Dir2,No)
% For a cell, return linear ind. of vert. in Node (Idx) and # in P (CellNo)
Rows = Row + [0 Dir2(No,1) Dir1(No,1)+Dir2(No,1) Dir1(No,1)]';
Cols = Col + [0 Dir2(No,2) Dir1(No,2)+Dir2(No,2) Dir1(No,2)]';
Idx = Rows + (Cols-1)*Grid(1) + ((1:3)-1)*Grid(1)*Grid(2);
CellNo = Rows(No) + (Cols(No)-1)*(Grid(1)-1);
end
function Vert = CellVertCoor(a,Vert,F,Ang)
% Calculte unknown vertices in cell depending on length of design var. a 
if length(a) == 1 % Step 1 (second node rel. to origin node)
    Vert(2,1:2) = Vert(1,1:2) + a*[cosd(Ang+90) sind(Ang+90)];
    Vert(2,3) = F(Vert(2,1),Vert(2,2));
elseif length(a) == 4 % Step 2 (Vert 3 rel. to 2 and Vert 4 rel. to 1) 
    Vert(3,:) = [Vert(2,1:2)+a(1:2) F(Vert(2,1)+a(1),Vert(2,2)+a(2))];
    Vert(4,:) = [Vert(1,1:2)+a(3:4) F(Vert(1,1)+a(3),Vert(1,2)+a(4))];
elseif length(a) == 2 % Step 3 (Vert 3 rel. to 2)
    Vert(3,:) = [Vert(2,1:2)+a(1:2) F(Vert(2,1)+a(1),Vert(2,2)+a(2))];
end
end
function [Out1, Out2] = DistFun(a,Vert,F,d,Ang)
% Get cell vertices and return distance between vertices depending on step
Vert = CellVertCoor(a,Vert,F,Ang);
if length(a) == 1 % Step 1 (1 x dist)
    Out1 = norm(Vert(2,:)-Vert(1,:)) - d;
elseif length(a) == 4 % Step 2 ([] (inequality constr.) and 3 x dist)
    Out1 = [];
    Out2 = vecnorm(Vert([3,4,1],:)-Vert([2,3,4],:),2,2)' - d;
elseif length(a) == 2 % Step 3 (2 x dist.)
    Out1 = vecnorm(Vert([3,4],:)-Vert([2,3],:),2,2)' - d;
end
end
function [Obj, Vert, Shear] = ShearFun(a,Vert,F)
% Get cell vertices and calc. shear angles. Return shear ang. sum (obj. in 
% step 2), vertex coordinates and vector of shear angles (for P array)
Vert = CellVertCoor(a,Vert,F);
% Calculate shear angles using cell edge vectors u and v
u = Vert([2 3 4 1],:)' - Vert'; 
v = Vert([4 1 2 3],:)' - Vert';
Shear = abs(atan2d(vecnorm(cross(u,v),2,1),dot(u,v))-90);
Obj = sum(Shear); 
end

File: Python implementation/KinDrape.py
import numpy as np 
from scipy.interpolate import CloughTocher2DInterpolator as CT2DInt
from scipy.optimize import (fsolve, minimize)
from mpl_toolkits.mplot3d import (Axes3D, art3d)
import matplotlib as mpl
import matplotlib.pyplot as plt
def KinDrape(d, Grid, Org, Ang, OrgNode):
    ## Mold definition: Hemisphere
    The, Phi = np.meshgrid(np.linspace(0,2*np.pi,100),
                           np.linspace(1e-6,np.pi/2-np.pi/20,50))
    X = np.cos(The)*np.sin(Phi); Y = np.sin(The)*np.sin(Phi); Z = np.cos(Phi)
    F = CT2DInt((X.ravel(),Y.ravel()),Z.ravel(),fill_value = np.min(Z))
    ## Aux. variables, solver settings and initialization of Node, P and Shear
    Node = np.empty((*Grid, 3))*np.nan
    P = [list(((np.nan,np.nan,np.nan),)*4)]*(Grid[0]-1)*(Grid[1]-1)
    Shear = np.empty(((Grid[0]-1)*(Grid[1]-1),4))*np.nan 
    Dir1 = np.array([[1, 0], [0, 1], [-1, 0], [0, -1]])  
    Dir2 = np.array([[0, 1], [-1, 0], [0, -1], [1, 0]]) 
    ## STEP 1: Place org. node (1) and node (2) defined by ini. drape angle
    # Get indices for cell, place 1st node and solve for 2nd node
    Idx = CellIdx(OrgNode[0],OrgNode[1],Dir1,Dir2,Grid,0)[0]
    Node[tuple(Idx[:,0])] = [Org[0], Org[1], F(Org[0], Org[1])]
    a_sol = fsolve(DistFun, d, args=(Node[tuple(Idx[:,0:2])], F, d, Ang))
    Node[tuple(Idx[:,0:2])] = CellVertCoor(a_sol,Node[tuple(Idx[:,0:2])],F,Ang)
    ## STEP 2: Place generator cells (initial cells) while minimizing shear
    GenStart = OrgNode + np.array([[0, 0], [1, 1], [0, 1], [0,0]])
    nGen = np.hstack([Grid[0]-OrgNode[0]-1, Grid[1]-OrgNode[1]-2,  OrgNode])
    Opt = {'maxiter': 100, 'ftol': 5e-9, 'iprint': 0, 'disp': True}
    for i in range(4):  # arms
        a_0 = 3/4*d*np.tile(CosSin(Ang+i*90),(1,2))
        for k in GenStart[i,:] + np.arange(nGen[i]).reshape(-1,1)*Dir1[i,:]:
            # Define idx and solver input. Call optimizer, assign solution
            Idx, CellNo = CellIdx(k[0],k[1],Dir1,Dir2,Grid,i) 
            BndBox = 0.5*np.array([[-1,-1,-1,-1],[1,1,1,1]])
            Bnd = tuple(map(tuple,(a_0+np.linalg.norm(a_0[0:3])*BndBox).T))
            Arg = (Node[tuple(Idx)], F, d)
            Constr = {'type': 'eq', 'fun': DistFun, 'args': Arg}
            a_sol = minimize(ObjFun, a_0, args=Arg, bounds=Bnd,
                           method='SLSQP', constraints=Constr, options=Opt)
            # Put current cell coord. and shear in P and Shear and update a_0
            Node[tuple(Idx)] = CellVertCoor(a_sol.x, Node[tuple(Idx)], F)
            P[CellNo] = list(map(tuple,Node[tuple(Idx)]))
            Shear[CellNo,:] = ShearFun(Node[tuple(Idx)])
            a_0 = a_sol.x     
    ## STEP 3: Place remaining, constrained cells in 4 quadrants between arms
    ConStart = OrgNode + np.array([[1,1],[0,1],[0,0],[1,0]])
    nCon = (nGen[[0,1,2,1,2,3,0,3]]-[1,0,0,0,0,0,1,0]).reshape(4,2)
    for i in range(4): 
        for h in ConStart[i,0] + np.arange(nCon[i,0])*(Dir1[i,0]+Dir2[i,0]):
            for w in ConStart[i,1] + np.arange(nCon[i,1])*(Dir1[i,1]+Dir2[i,1]):  
                # Define idx and solver input. Call fsolve, assign solution
                Idx, CellNo = CellIdx(h,w,Dir1,Dir2,Grid,i) 
                a_0 = Node[tuple(Idx[:,3])][0:2] - Node[tuple(Idx[:,0])][0:2] 
                a_sol = fsolve(DistFun, a_0, args=(Node[tuple(Idx)], F, d),
                                factor=1,diag=(d,d))
                Node[tuple(Idx)] = CellVertCoor(a_sol, Node[tuple(Idx)], F)
                # Put curr. cell coord. and shear in P and Shear and upd. a_0
                P[CellNo] = list(map(tuple,Node[tuple(Idx)]))
                Shear[CellNo,:] = ShearFun(Node[tuple(Idx)])
    ## Plotting
    # Create 3D figure and plot surface
    fig = plt.figure(); ax = Axes3D(fig)
    ax.plot_surface(X,Y,Z,rstride=2,cstride=2,color=(0.64,0.71,0.8),shade=True)
    # Define colormap and map the mean shear of each cell to a list of colors
    cMin = np.nanmin(Shear); cMax = np.nanmax(Shear);  
    CMapName = 'jet'; CMap = mpl.cm.get_cmap(CMapName);
    C = list(map(list,CMap(mpl.colors.Normalize(cMin,cMax)(np.mean(Shear,1)))))
    # Plot cells as colored polygons, and create axis labels and colorbar
    pc = art3d.Poly3DCollection(P,cmap=CMapName)
    pc.set_facecolor(C); pc.set_edgecolor('k'); ax.add_collection3d(pc)
    ax.set_xlabel('x'); ax.set_ylabel('y'); ax.set_zlabel('z'); 
    ax.set_box_aspect((np.ptp(X), np.ptp(Y), np.ptp(Z)))
    fig.colorbar(pc,shrink=0.75,boundaries=np.linspace(cMin, cMax, 50),
                  label='Shear angle [deg]')
    plt.show() 
    return Node, Shear
## Auxiliary functions
def CosSin(Alpha):
    # Return a unit vector in direction of Alpha
    return np.array([np.cos(Alpha*np.pi/180), np.sin(Alpha*np.pi/180)])
def CellIdx(Row,Col,Dir1,Dir2,Grid,No):
    # Return all row and col indices for the cell given vertex 1
    Rows = Row + np.array([0, Dir2[No,0], Dir1[No,0]+Dir2[No,0], Dir1[No,0]])
    Cols = Col + np.array([0, Dir2[No,1], Dir1[No,1]+Dir2[No,1], Dir1[No,1]])
    CellNo = Rows[No] + Cols[No]*(Grid[0]-1);
    return np.vstack([Rows, Cols]), CellNo
def DistFun(a, Vert, F, d, Ang=0.0):  
    # Return the difference between the edge lengths and discretization d
    Vert = CellVertCoor(a, Vert, F, Ang)
    if a.size == 4:
        EdgeLength = np.linalg.norm(Vert[[2,3,0],:] - Vert[[1,2,3],:],2,1)
    elif a.size == 2:
        EdgeLength = np.linalg.norm(Vert[[2,3],:] - Vert[[1,2],:],2,1)
    elif a.size == 1:
        EdgeLength = np.linalg.norm(Vert[0,:] - Vert[1,:])
    return EdgeLength-d  
def ShearFun(Vert):  
    # Calculate shear angles in the cell using edge vector pairs u and v
    u = Vert[[1, 2, 3, 0], :] - Vert
    v = Vert[[3, 0, 1, 2], :] - Vert
    CellAng = np.arctan2(np.linalg.norm(np.cross(u,v),2,1),np.sum(u*v,1))
    return np.abs(CellAng*180/np.pi - 90)
def ObjFun(a, Vert, F, d):
    # The objectice function from STEP 2 as sum of shear angles
    Vert = CellVertCoor(a, Vert, F)
    return np.sum(ShearFun(Vert))
def CellVertCoor(a, Vert, F, Ang=0.0):  
    # Calculate vertices in a cell given known vertices and design vars.
    if a.size == 1: # STEP 1
        Vert[1,0:2] = Vert[0,0:2]+a*CosSin(Ang+90)
        Vert[1,2] = F(Vert[1, 0], Vert[1, 1]) 
    if a.size >= 2: # STEP 2 and 3
        Vert[2,0:2] = Vert[1,0:2] + a[0:2]
        Vert[2,2] = F(Vert[1,0] + a[0],Vert[1,1] + a[1])
    if a.size == 4: # STEP 2
        Vert[3,0:2] = Vert[0,0:2] + a[2:4]
        Vert[3,2] = F(Vert[0,0] + a[2],Vert[0,1] + a[3])
    assert ~np.any(np.isnan(Vert)), 'Interpolant F is evaluated outside domain'
    return Vert
